LFX'25: Pkg Telemetry: Backend Otel Instrumentation

DhairyaMajmudar opened the pull request:
## Tasks Done

1. Installing required Otel packages for instrumenting backend.
2. Developed a new package named `telemetry` where otel exporters, providers, and middlewares are present.
4. Added test cases for every new function of the `telemetry` package.

## How to test

Run the command `cd backend/pkg/telemetry && go test -v . . `

This command will trigger the unit test cases for the `telemetry` package.

## Demo video

[Screencast from 2025-04-10 15-03-59.webm](https://github.com/user-attachments/assets/2b73e15d-2064-43ea-bacc-3210db55afe0)

## Reference documentation 

1. https://opentelemetry.io/docs/languages/go/getting-started/
2. https://opentelemetry.io/docs/languages/go/instrumentation/
3. https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric@v1.35.0
4. https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.35.0/trace

Related to https://github.com/headlamp-k8s/headlamp/issues/2799

illume commented:
btw, to see the lint issues locally with `make backend-lint`

illume commented:
Regarding the fossa License Compliance failing check, I'm not sure what that is. It seems all the packages you've added are Apache 2.0.

illume commented:
@DhairyaMajmudar can you please add into the PR description a section on "How to test"?

illume commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,272 @@
+)
+
+type Config struct {
+	ServiceName        string
```
Can you please document these fields?

illume commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+type Telemetry struct {
+	config         Config
```
Please document these fields?

illume commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+type Metrics struct {
+	RequestCounter       metric.Int64Counter
```
Please document these fields?

illume commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -0,0 +1,151 @@
+	statusCode int
+}
+
+func newResponseWriter(w http.ResponseWriter) *responseWriter {
```
This one is missing documentation.

illume commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,272 @@
+	shutdown       func(context.Context) error
+}
+
+func NewTelemetry(cfg Config) (*Telemetry, error) {
```
This one is missing documentation.

illume commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,272 @@
+	return nil
+}
+
+func setupShutdownFunction(t *Telemetry) {
```
This one is missing documentation.

illume commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,272 @@
+	}
+}
+
+func (t *Telemetry) Shutdown(ctx context.Context) error {
```
This one is missing documentation.

illume commented on backend/pkg/telemetry/tracing.go:
```diff
@@ -0,0 +1,51 @@
+	"go.opentelemetry.io/otel/trace"
+)
+
+func TracingMiddleware(serviceName string) func(http.Handler) http.Handler {
```
This one is missing documentation.

illume commented on backend/cmd/headlamp.go:
```diff
@@ -929,7 +932,49 @@ func (c *HeadlampConfig) OIDCTokenRefreshMiddleware(next http.Handler) http.Hand
 }
 
+func initializeTelemetry(config *HeadlampConfig) {
+	telemetryCfg := telemetry.DefaultConfig()
```
Would it make sense if this was moved within the pkg/telemetry package? Then it could be called from here.

illume commented on backend/cmd/headlamp.go:
```diff
@@ -929,7 +932,49 @@ func (c *HeadlampConfig) OIDCTokenRefreshMiddleware(next http.Handler) http.Hand
 	})
 }
 
+func initializeTelemetry(config *HeadlampConfig) {
```
This one is missing docs.

illume reviewed (changes requested):
Thanks for this.

Adding a "How to test" section to the PR would be helpful so that people can manually try it out.

Would you mind adding in comments for each file a couple of the main documentation links that are relevant to it? The ones that you read whilst making it would be good.

It could be improved to break it up into smaller atomic commits. One for each function would be best (or more than one if they are tightly coupled of course!).

When you add a function, it would be good to add a test for it in the same commit.

DhairyaMajmudar commented:
- [x] Added "How to test" section in PR description.
- [x] Segregated commits into atomic and unique for each function.
- [x] Added documentation comments for the function.
- [x] Added test cases for each and every function of `pkg/telemetry`.

illume commented:
Very cool.

Would you mind adding the dependencies in separate commits?

Are the ones updating existing dependencies required for this PR?

DhairyaMajmudar commented:
Tests are running in my local, but they are still failing in the CI. Is this because we've moved our repo. to a new location ?

illume commented:
@DhairyaMajmudar there's a backend test failing in main. Yes related to the repo move. Hopefully someone will fix the error in main shortly.

illume commented on backend/go.mod:
```diff
@@ -140,17 +153,16 @@ require (
-	golang.org/x/crypto v0.35.0 // indirect
-	golang.org/x/net v0.36.0 // indirect
-	golang.org/x/sync v0.11.0 // indirect
+	golang.org/x/crypto v0.32.0 // indirect
```
Can you check this? I'm not sure why these libraries are being downgraded to older versions.

illume commented:
Is there a way to test this to see the output somehow? If so, can you add instructions of that to the PR in the how to test section?

DhairyaMajmudar commented:
> Is there a way to test this to see the output somehow? If so, can you add instructions of that to the PR in the how to test section?

To check the outputs I need to add the `initTelemetry` function in headlamp.go file. I've did this in my local let me add small demo video for the same.

illume commented:
Can you please remove the merge main commits?
(Instead you can rebase against a fresh main)

illume reviewed (changes requested):
Hello. I left some notes on each commit.

If you go to the commit tab, and go through one commit at a time, it should make more sense which comment is related to which commit.

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -300,3 +300,19 @@ func (t *Telemetry) Shutdown(ctx context.Context) error {
+// DefaultConfig provides a reasonable default configuration for telemetry.
+// It enables both tracing and metrics with 100% sampling rate and
+// configures exporters to local endpoints for development use.
+func DefaultConfig() Config {
```
Nope, these default config is configured for production mode,

For the developers running and testing the code in development mode I would do something like this in `headlamp.go` file to change some of these configs 

```golang
if config.devMode {
    telemetryCfg.StdoutTraceEnabled = true
 }
```

This will enable spans to be printed in the terminal in dev mode.

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -300,3 +300,19 @@ func (t *Telemetry) Shutdown(ctx context.Context) error {
+func DefaultConfig() Config {
+	return Config{
+		ServiceName:        "headlamp",
+		ServiceVersion:     "1.0.0",
```
I feel this should be kept same as app/pacakge.json because in public headlamp releases, also we mention the same version 

see: https://github.com/kubernetes-sigs/headlamp/releases/tag/v0.30.0

illume replied:
Can you think of a way so we don't have to maintain this manually every release?

DhairyaMajmudar replied:
Very valid point, let me look for possible ways for this

DhairyaMajmudar replied:
Can we do something like this in makefile? 

```
APP_VERSION := $(shell grep -o '"version": "[^"]*"' app/package.json | cut -d '"' -f 4)

build:
    go build -ldflags "-X github.com/kubernetes-sigs/headlamp/backend/pkg/telemetry.appVersion=$(APP_VERSION)" ./...

```

Extracting the version from package.json and passing it in build step.

illume replied:
We decided to leave this out of this PR. The Makefile needs to work without unix tools, like when running on windows or in a very limited container.

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -73,6 +73,45 @@ func createResource(cfg Config) (*resource.Resource, error) {
 	return res, nil
 }
 
+// createTracingExporter creates an appropriate span exporter based on configuration.
```
Such better changed with the suggested.

DhairyaMajmudar replied:
Yeah! Good one, replaced with the suggested line.

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -73,6 +73,45 @@ func createResource(cfg Config) (*resource.Resource, error) {
 
+// createTracingExporter creates an appropriate span exporter based on configuration.
+// It prioritizes exporters in the following order:
+// 1. Stdout exporter (if StdoutTraceEnabled is true)
```
Added

DhairyaMajmudar replied:
Added 👍🏻

DhairyaMajmudar commented on backend/pkg/telemetry/tracing_test.go:
```diff
@@ -78,6 +79,56 @@ func TestTracingMiddleware(t *testing.T) {
 	assert.True(t, containsAttribute("http.status_code", 200))
 }
 
+func TestTracingMiddlewareWithPropagation(t *testing.T) {
```
Yeah! Moved the test cases in that commit

DhairyaMajmudar commented on backend/pkg/telemetry/tracing_test.go:
```diff
@@ -153,3 +153,45 @@ func TestStartSpan(t *testing.T) {
 	assert.True(t, found, "Expected to find attribute test.key=test-value")
 }
+
+func TestAddSpanAttributes(t *testing.T) {
```
Moved the test to its implemenation commit : )

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry_test.go:
```diff
@@ -224,3 +225,33 @@ func TestCreateSampler(t *testing.T) {
+	_, span := tracer.Start(context.Background(), "test-span")
+	span.End()
+
+	// ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
I commented out that piece of code because it required the function `Shutdown` which will be further add in the next commit  https://github.com/kubernetes-sigs/headlamp/pull/2976/commits/0353b7a4190b39482788aee04271cee1df23d09c

DhairyaMajmudar commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -106,22 +125,32 @@ func NewMetrics() (*Metrics, error) {
+// 2. Records the request count with method and path attributes
+// 3. Captures the response status code
+// 4. Decrements the active requests gauge when the request completes
+func (m *Metrics) RequestCounterMiddleware(next http.Handler) http.Handler {
```
Let's do this in way.

illume replied:
Ok, if there’s no reason to do it like the first way then let’s keep it doing it all in one step.

I thought perhaps there could be timing differences with the first way?

DhairyaMajmudar replied:
The first approach which I implemented were collecting metrics in two steps: 1) Request start 2) Request completion.

DhairyaMajmudar commented on backend/pkg/telemetry/metrics_test.go:
```diff
@@ -126,3 +126,87 @@ func TestNewMetrics(t *testing.T) {
 	}
 	assert.True(t, found, "Expected to find http.server.request_count metric")
 }
+
```
Done, added test cases to check panic handling : )

DhairyaMajmudar commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -106,22 +125,32 @@ func NewMetrics() (*Metrics, error) {
+
+		wrapper := newResponseWriter(w)
+
+		next.ServeHTTP(wrapper, r)
```
Added logic to record panic handling, it will record the status code as 500, I've also added tests cases to verify the same

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -73,6 +73,45 @@ func createResource(cfg Config) (*resource.Resource, error) {
+// 2. OTLP exporter (if OTLPEndpoint is set)
+// 3. Defaults to stdout exporter if no other options are available
+func createTracingExporter(cfg Config) (trace.SpanExporter, error) {
+	if cfg.StdoutTraceEnabled {
```
Done added the warnings logic

DhairyaMajmudar commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -196,6 +196,28 @@ func createStdoutExporter() (trace.SpanExporter, error) {
+			err2 = t.meterProvider.Shutdown(ctx)
+		}
+
+		if err1 != nil {
```
Fixed this one, it will onw have logic to handle multiple errors

DhairyaMajmudar commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -63,7 +65,7 @@ type Metrics struct {
 // active request monitoring, cluster proxy usage, plugin loading, and error counting.
 // The returned metrics instance can be used throughout the application to record metrics data.
-func NewMetrics() (*Metrics, error) {
+func NewMetrics() (*Metrics, error) { //nolint:funlen
```
Done, I segregated function, into two new functions

DhairyaMajmudar replied:
It can be found in the commit https://github.com/kubernetes-sigs/headlamp/pull/2976/commits/df33b0528e963650e466166dc5286fe5a7ff1e56#diff-74fe0982830645e9ade05b76518f6925262271e6f27f00dd78f26ca1d7ddaa13R88

illume commented:
There’s a commit which references the wrong issue. See message above.

illume commented on backend/go.mod:
```diff
@@ -32,7 +32,19 @@ require (
 	github.com/coreos/go-oidc/v3 v3.11.0
 	github.com/gorilla/websocket v1.5.3
-	golang.org/x/term v0.29.0
+	github.com/prometheus/client_golang v1.19.1
```
Can you please do this one? It will make debugging later easier, and be helpful for people trying to understand the code.

DhairyaMajmudar replied:
I tried to do this, but it seems to be a bit tricky as I've to make edits to all the commits by doing a lot of rebasing stuff. But yeah you're correct too. It will help in debugging a lot. Let me give it an another try.

DhairyaMajmudar replied:
Done, this time it worked. We are now good to go.

illume reviewed (commented):
There’s two notes that are unresolved. Can you please have a look at them?

The dependencies being separated into commits is the main one left.

The other one is left open has a question about the differences between the two approaches. Is there any differences functionally between the two approaches?

DhairyaMajmudar commented:
> There’s two notes that are unresolved. Can you please have a look at them?

Done, I've answered both of them.

> There’s also a problem with one of the commit messages that seems to include a bunch of old messages for some reason. See link above.

It is now fixed, it seems like it caused while doing rebase to main.

> btw. We don’t need signed off by in the commit messages anymore. Since the k8s project uses the CLA thing.

That's cool, thank you for letting me know : )

DhairyaMajmudar commented:
>   Include a link to the original issue in the PR description.

I've added the link for the parent issue https://github.com/kubernetes-sigs/headlamp/issues/2799 in PR description 


>   Include in the PR description the related documentation that you used for writing this? This could also be placed somewhere in a docs/backend folder or in a telementry/README.md file.

Done I've updated the PR with the relevant docs & also added a README.md file in the telemetry package 

>   Write in text the "How to test" instructions for testing with Jaeger? ( I know the video sort of shows this). Perhaps, including a link to the jaeger quickstart documentation (and a link to the docs for running the UI). Having the steps as text will make it easy for people to test.

For running up the jaeger demonstration one needs the instrumented code of API routes. As this PR is focused on adding the core telemetry package (Which would be further useful for API route instrumentation) I've not added that in the same PR (I've planning to open an another PR for the same). Shall I make the docs/readme updation in the PR?

illume commented:
> For running up the jaeger demonstration one needs the instrumented code of API routes. As this PR is focused on adding the core telemetry package (Which would be further useful for API route instrumentation) I've not added that in the same PR (I've planning to open an another PR for the same). Shall I make the docs/readme updation in the PR?

Sounds good.

illume reviewed (approved):
This looks good to me.

@yolossn @knrt10 Please review?

yolossn commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,351 @@
+	MetricsEnabled bool
+	// JaegerEndpoint is the endpoint for Jaeger tracing backend
+	// Setting this will also set OTLPEndpoint if it's not already set
+	JaegerEndpoint string
+	// OTLPEndpoint is the endpoint for OTLP-compatible tracing backends
+	OTLPEndpoint string
```
Use pointers for optional configs

DhairyaMajmudar replied:
on it and I can also see the `promethusPort` would not be useful further more as I've removed the metric server to the main headlamp server. Let me remove that too.

yolossn commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,351 @@
+
+// Config defines the configuration options for telemetry initialization.
+// It controls how tracing and metrics are collected and where they are exported.
+type Config struct {
+	// ServiceName is the name of the service being monitored (required)
+	ServiceName string
+	// ServiceVersion is the version of the service being monitored
+	ServiceVersion string
+	// TracingEnabled determines if distributed tracing is enabled
+	TracingEnabled bool
+	// MetricsEnabled determines if metrics collection is enabled
+	MetricsEnabled bool
+	// JaegerEndpoint is the endpoint for Jaeger tracing backend
+	// Setting this will also set OTLPEndpoint if it's not already set
+	JaegerEndpoint string
+	// OTLPEndpoint is the endpoint for OTLP-compatible tracing backends
+	OTLPEndpoint string
```
IMHO the config should be part of [config](https://github.com/kubernetes-sigs/headlamp/blob/main/backend/pkg/config/config.go) package for centralised config management.

DhairyaMajmudar replied:
Awsm! Let me transfer this and its corresponding test cases to that file, Thank you!

DhairyaMajmudar replied:
Done in my latest commit https://github.com/kubernetes-sigs/headlamp/pull/2976/commits/652e62b5421555def0f20512944882b029c90936

1. I've converted optional configs into pointers
2. Moved config struct into config package (Shall I also moved `defaultConfig` function to that? POV: It should be reminded here in the telemetry package as it is related to that, but yeah would love to know your opinion on this)
3. Removed test cases which was checking internal private functions from telemetry_test file.
4. I also added a `//nolint:testpackage` comment in the test file to which I've removed too.

yolossn commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,351 @@
+// DefaultConfig provides a reasonable default configuration for telemetry.
+// It enables both tracing and metrics with 100% sampling rate and
+// configures exporters to local endpoints for development use.
+func DefaultConfig() Config {
```
I don't see the `DefaultConfig` function being used anywhere.

yolossn commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -0,0 +1,178 @@
+	// ActiveRequestsGauge tracks the number of currently active HTTP requests
+	ActiveRequestsGauge metric.Int64UpDownCounter
+	// ClusterProxyRequests counts requests made through the cluster proxy
+	ClusterProxyRequests metric.Int64Counter
```
`ClusterProxyRequests` counter is only defined and initialised, I cant see it being used anywhere.

DhairyaMajmudar replied:
Thank you for the review comment the `ClusterProxyRequests` would be further used in the route instrumentation of `headlamp.go` file (I'm planning to open a new corresponding PR for route instrumentation). Below is the code snippet for the same

```golang
if c.metrics != nil {
				c.metrics.ClusterProxyRequests.Add(ctx, 1,
					metric.WithAttributes(
						attribute.String("cluster", contextKey),
						attribute.String("http.method", r.Method),
					))
			}
```

yolossn replied:
AFAIU you can do this in the middleware itself. You can find which route the request is being routed to and instrument the metric.


Please refer this example below

```go
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/gorilla/mux"
)

// Middleware that logs the matched route path template
func LoggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		route := mux.CurrentRoute(r)
		if route != nil {
			if pathTemplate, err := route.GetPathTemplate(); err == nil {
				log.Printf("Matched route: %s", pathTemplate)
			}
		}

		vars := mux.Vars(r)
		if vars["clusterName"] != "" {
			log.Printf("Cluster: %s", vars["clusterName"])
			// TODO: ADD logic for cluster metrics handling
		}

		next.ServeHTTP(w, r)
	})
}

// Handler for cluster API requests
func clusterHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	fmt.Fprintf(w, "Cluster: %s, API Path: %s\n", vars["clusterName"], vars["api"])
}

func main() {
	r := mux.NewRouter()
	r.Use(LoggingMiddleware) // Add middleware

	r.PathPrefix("/clusters/{clusterName}/{api:.*}").HandlerFunc(clusterHandler)

	log.Println("Server started at http://localhost:8080")
	log.Fatal(http.ListenAndServe(":8080", r))
}
```

DhairyaMajmudar replied:
I don't feel it will be a good idea to add that into the middleware functions because in the `headlamp.go` file there are several other instances such as below, and it would be good to keep that logic in `headlamp.go` file.

```golang
c.metrics.RequestCounter.Add(ctx, 1,
	metric.WithAttributes(attribute.String("api.route", "/cluster/delete")))
```

DhairyaMajmudar commented on backend/pkg/telemetry/metrics.go:
```diff
@@ -0,0 +1,178 @@
+	// ClusterProxyRequests counts requests made through the cluster proxy
+	ClusterProxyRequests metric.Int64Counter
+	// PluginLoadCount tracks the number of plugin loads
+	PluginLoadCount metric.Int64Counter
+	// PluginDeleteCount tracks the number of plugin deletions
+	PluginDeleteCount metric.Int64Counter
```
Same goes with these too, they would be further useful in `headlamp.go` file : )

yolossn replied:
Okay, Please add kubeconfig refresh counter to the list of metrics. Feel free to do this in the subsequent PR also.

DhairyaMajmudar replied:
I've added this in present PR

illume commented:
Thanks for the review @yolossn and for addressing those points @DhairyaMajmudar

Let's leave it up to @yolossn to mark the items resolved, so they can review the changes more easily.

yolossn commented on backend/pkg/config/config.go:
```diff
@@ -41,6 +41,17 @@ type Config struct {
 	OidcValidatorIdpIssuerURL string `koanf:"oidc-validator-idp-issuer-url"`
 	OidcScopes                string `koanf:"oidc-scopes"`
 	OidcUseAccessToken        bool   `koanf:"oidc-use-access-token"`
+
```
Please add corresponding flags to the `flagset` function in the same file, also if required add validations for these fields like if tracing is enabled atleast one of the `JaegerEndpoint`, `OTLPEndpoint` should be configured etc.

DhairyaMajmudar replied:
Done

yolossn commented on backend/pkg/config/config.go:
```diff
@@ -41,6 +41,17 @@ type Config struct {
 	OidcScopes                string `koanf:"oidc-scopes"`
 	OidcUseAccessToken        bool   `koanf:"oidc-use-access-token"`
+
+	// telemetry configs
+	ServiceName        string
+	ServiceVersion     *string
+	TracingEnabled     *bool
+	MetricsEnabled     *bool
+	JaegerEndpoint     *string
+	OTLPEndpoint       *string
+	UseOTLPHTTP        *bool
+	StdoutTraceEnabled *bool
+	SamplingRate       *float64
```
add `koanf` reflections for these configs

DhairyaMajmudar replied:
Added

yolossn commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,331 @@
+	version    = "0.30.0"
+	trueVal    = true
+	falseVal   = false
+	endpoint   = "localhost:4317"
```
looks like this value is not used, please remove it.

DhairyaMajmudar replied:
Removed that whole and added to the flags

yolossn commented on backend/pkg/telemetry/telemetry.go:
```diff
@@ -0,0 +1,331 @@
+// DefaultConfig provides a reasonable default configuration for telemetry.
+// It enables both tracing and metrics with 100% sampling rate and
+// configures exporters to local endpoints for development use.
+func DefaultConfig() cfg.Config {
```
The default values can be configured in the `flagset()` function in the config package

https://github.com/kubernetes-sigs/headlamp/blob/06eeaf6774c2aafb3fc7fd5028e65d530afff23b/backend/pkg/config/config.go#L169

DhairyaMajmudar replied:
That's a very good idea, I've now removed the `DeafultConfig` function and added them to the flags

DhairyaMajmudar commented:
Thank you @yolossn for the review comments, I've made the suggested changes : )
